From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <nonplay@bxteam.org>
Date: Sun, 6 Jul 2025 01:55:37 +0300
Subject: [PATCH] C2ME: Optimize world gen math

This patch is based on following mixins:
* "com/ishland/c2me/opts/math/mixin/MixinChunkPos.java"
* "com/ishland/c2me/opts/worldgen/vanilla/mixin/structure_weight_sampler/MixinStructureWeightSampler.java"
By: ishland <ishlandmc@yeah.net>
As part of: C2ME-fabric (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT

TODO: Drop NoiseBasedChunkGenerator, move to single patch

diff --git a/net/minecraft/world/level/levelgen/Beardifier.java b/net/minecraft/world/level/levelgen/Beardifier.java
index 6781da6102df878da9db986f2687fbc47efcc8be..e9b049e0c1f41962d481c44c568a92f21ffa1c51 100644
--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -166,8 +166,14 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
     }
 
     private static double getBuryContribution(final double dx, final double dy, final double dz) {
-        double distance = Mth.length(dx, dy, dz);
-        return Mth.clampedMap(distance, 0.0, 6.0, 1.0, 0.0);
+        // DivineMC start - C2ME: Optimize world gen math
+        double length = Math.sqrt(dx * dx + dy * dy + dz * dz);
+        if (length > 6.0) {
+            return 0.0;
+        } else {
+            return 1.0 - length / 6.0;
+        }
+        // DivineMC end - C2ME: Optimize world gen math
     }
 
     private static double getBeardContribution(final int dx, final int dy, final int dz, final int yToGround) {
diff --git a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
index 83bc0f6fa7624a0df546726fa8ccc9744173ff88..8d0c3f96097c42a257292ab27d9bae18f66d9bdf 100644
--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -75,7 +75,10 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
             if (SharedConstants.DEBUG_DISABLE_FLUID_GENERATION) {
                 return emptyStatus;
             } else {
-                return y < Math.min(-54, seaLevel) ? lavaStatus : seaStatus;
+                // DivineMC start - C2ME: Optimize world gen math
+                final int min = Math.min(-54, seaLevel);
+                return y < min ? lavaStatus : seaStatus;
+                // DivineMC end - C2ME: Optimize world gen math
             }
         };
     }
